src/systems/buoyancy/Buoyancy.cc

```diff
@@ -54,37 +56,147 @@ using namespace systems;
+  /// link.
+  public:
+  template<typename T>
+  void GradedFluidDensity(
+    const math::Pose3d &_pose, const T &_shape, const math::Vector3d _gravity);
 
   /// \brief Model interface
   public: Entity world{kNullEntity};
 
   /// \brief The density of the fluid in which the object is submerged in
   /// kg/m^3. Defaults to 1000, the fluid density of water.
   public: double fluidDensity{1000};
+
+  /// \brief When using GradedBuoyancy, we provide a different buoyancy for
+  /// each layer. The key on this map is height in meters and the value is fluid
+  /// density. I.E all the fluid between $key$m and $next_key$m has the density
+  /// $value$kg/m^3. Everything below the first key is considered as having
+  /// fluidDensity.
+  public: std::map<double, double> layers;
+
+  public: struct BuoyancyActionPoint
+  {
+    math::Vector3d force;
+    math::Vector3d point;
+    math::Pose3d pose;
+  };
+
+  public: std::vector<BuoyancyActionPoint> buoyancyForces;
+
+  public: std::pair<math::Vector3d, math::Vector3d> resolveForces(
+    const math::Pose3d &_pose);
 };
 
 //////////////////////////////////////////////////
-double BuoyancyPrivate::FluidDensity(const math::Pose3d & /*_pose*/) const
+double BuoyancyPrivate::UniformFluidDensity(const math::Pose3d &/*_pose*/) const
 {
-  // \todo(nkoenig) Adjust the fluid density based on the provided pose.
-  // This could take into acount:
-  //   1. Transition from water to air. Currently this function is used for
-  //   a whole link, but when transitioning between mediums a link may span
-  //   both mediums. Surface tension could also be a factor.
-  //   2. Fluid density changes based on depth below the water surface and
-  //   height above water surface.
   return this->fluidDensity;
 }
 
+//////////////////////////////////////////////////
+template<typename T>
+void BuoyancyPrivate::GradedFluidDensity(
+  const math::Pose3d &_pose, const T &_shape, const math::Vector3d _gravity)
+{
+  auto prevLayerFluidDensity = this->fluidDensity;
+  auto volsum = 0.0;
+  auto centerOfBuoyancy = math::Vector3d{0, 0, 0};
+
+  for(auto [height, currFluidDensity] : layers)
+  {
+    // TODO(arjo): Transform plane and slice the shape
+    math::Planed plane{math::Vector3d{0, 0, 1}, height - _pose.Pos().Z()};
+    auto vol = _shape.VolumeBelow(plane);
+
+    // Archimedes principal for this layer
+    auto forceMag =  - (vol - volsum) * _gravity * prevLayerFluidDensity;
+
+    // Accumulate layers.
+    prevLayerFluidDensity = currFluidDensity;
+
+    // Calculate point from which force is applied
+    auto cov = _shape.CenterOfVolumeBelow(plane);
+
+    if(!cov.has_value()) continue;
+
+    auto cob = (cov.value() * vol - centerOfBuoyancy * volsum) / (vol - volsum);
+    centerOfBuoyancy = cov.value();
+    auto buoyancyAction = BuoyancyActionPoint
+    {
+      forceMag,
+      cob,
+      _pose
+    };
+    buoyancyForces.push_back(buoyancyAction);
+
+    volsum = vol;
```

Comment:
Nope because VolumeBelowPlane returns all the volume below the current plane only. Hence the total volume seen until this point is still `vol`. I can rename this to prevent confusion.